python_modules/dagster/docs/intro_tutorial/types.rst

```diff
@@ -0,0 +1,233 @@
+User-Defined Types
+------------------
+
+Throughout the the tutorial you have seen the use of builtins such as `Int` and `String` for types.
+However you will want to be able to define your own dagster types to fully utilize the system. We'll
+go over that here.
+
+As a pedogogical example, we will go over the code used to the define the PandasDataFrame type in the
+dagster-pandas library, building it step by step along the way.
+
+Basic Typing
+^^^^^^^^^^^^
+
+.. code-block:: py
+
+    import pandas as pd
+    from dagster import as_dagster_type
+
+    DataFrame = as_dagster_type(
+        pd.DataFrame,
+        name='PandasDataFrame',
+        description='''Two-dimensional size-mutable, potentially
+    heterogeneous tabular data structure with labeled axes 
+    (rows and columns). See http://pandas.pydata.org/'''
+    )
+
+What this code doing is annotating/registering an existing type as a dagster type. Now one can include
+this type and use it as an input or output of a solid. The system will do a typecheck to ensure
+that the object is of type `pd.DataFrame`.
+
+Now one can use it to define a solid:
+
+.. code-block:: py
+
+    @lambda_solid(
+        inputs=[InputDefinition('num', dagster_pd.DataFrame)],
+        output=OutputDefinition(dagster_pd.DataFrame),
+    )
+    def sum_solid(num):
+        sum_df = num.copy()
+        sum_df['sum'] = sum_df['num1'] + sum_df['num2']
+        return sum_df
+
+The type metadata now appears in dagit and the system will ensure the input and output to this
+solid are indeed data frames.
+
+Input Schema
+^^^^^^^^^^^^
+
+This solid as defined is only expressed in terms of an in-memory object; it says nothing about
+how this data should be sourced from or materialized to disk. This is where the notion of
+input and output schemas come into play. Once the user provides those she is able to use
+the configuration language in order to parameterize the computation.
+
+Let us now add the input schema:
+
+.. code-block:: py
+
+    def define_path_dict_field():
+        return Field(Dict({'path': Field(Path)}))
+
+    def define_csv_dict_field():
+        return Field(
+            Dict(
+                {
+                    'path': Field(Path), 
+                    'sep': Field(
+                        String,
+                        is_optional=True,
+                        default_value=','
+                    )
+                }
+            )
+        )
+
+    @input_schema(
+        Selector(
+            {
+                'csv': define_csv_dict_field(),
+                'parquet': define_path_dict_field(),
+                'table': define_path_dict_field(),
+            }
+        )
+    )
+    def dataframe_input_schema(config_value):
+        # config_value is dictionary with a single element
```

Comment:
Nit: config_value doesn't *always* have a single element, right? That's only if the input_schema is a selector, right? This _kind_ of makes it seem like this is always true about config_value? I'm not sure how to make this comment better without making it more verbose--maybe add it to the text below this code snippet explaining it in more detail?